### PR TITLE
substitute using SymbolicUtils

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes https://github.com/SciML/ModelingToolkit.jl/issues/339

MWE:

```julia
using ModelingToolkit
@parameters α β γ δ
s = α => γ
expr = (((1 / β - 1) + δ) / α) ^ (1 / (α - 1))
ModelingToolkit.substitute(expr, s)
new  = (((1 / β - 1) + δ) / γ) ^ (1 / (γ - 1))
isequal(expr, new)
```

### CODE REVIEW

File: src/direct.jl [hunk 1]
```diff
@@ -34,7 +34,7 @@ an array of variable expressions.
 """
 function hessian(O::Expression, vars::AbstractVector{<:Expression}; simplify = true)
     out = [expand_derivatives(Differential(v2)(Differential(v1)(O))) for v1 in vars, v2 in vars]
-    simplify ? simplify_constants.(out) : out
+    simplify ? simplify.(out) : out
```
Comment (shashi) [inline src/direct.jl:37]: Don't think this will work :)

File: src/simplify.jl [hunk 2]
```diff
@@ -38,10 +35,12 @@ SymbolicUtils.symtype(x::Expression) = Number
 
 # SymbolicUtils -> ModelingToolkit
 
-function simplify_constants(expr)
+function SymbolicUtils.simplify(expr::Expression)
```
Comment (shashi) [inline src/simplify.jl:38]: Nice!

File: src/direct.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/direct.jl:37]: why not?

**Comment by ChrisRackauckas** [comment]
@isaacsas this'll handle the maj eval

**Comment by shashi** [comment]
https://github.com/JuliaRegistries/General/pull/14285 has and `substitute` that gets used here. We can add more smarts/perf improvements over at SymUtils.

File: Project.toml [hunk 3]
```diff
@@ -37,7 +37,7 @@ NaNMath = "0.3"
 SafeTestsets = "0.0.1"
 SpecialFunctions = "0.7, 0.8, 0.9, 0.10"
 StaticArrays = "0.10, 0.11, 0.12"
-SymbolicUtils = "0.1.1, 0.2"
+SymbolicUtils = "0.2.1, 0.2"
```
Comment (shashi) [inline Project.toml:40]: not sure if I did this right, so weird travis has no logs for the failure.

**Comment by ChrisRackauckas** [comment]
@shashi is this supposed to work with SymbolicUtils master?

```julia
using ModelingToolkit
@parameters α β γ δ
s = α => γ
expr = (((1 / β - 1) + δ) / α) ^ (1 / (α - 1))
ModelingToolkit.substitute(expr, s)
new  = (((1 / β - 1) + δ) / γ) ^ (1 / (γ - 1))
isequal(expr, new)
```

still fails for me.

**Comment by shashi** [comment]
No that's not supposed to work... `isequal` only checks the structure, it doesn't know that the two names are equal. That would require contextual simplification something like:

`simplify(expr == new, Constraints([α == γ]))`

**Comment by isaacsas** [comment]
@ChrisRackauckas @shashi This PR seems to have broken MassActionJumps. It appears test passed as I forgot to add the jump tests to the runtest file... Sorry about that.

There seem to be two issues: first the number of arguments in the call to `assemble_maj` is wrong as `ptoid` was not deleted from the function signature. Even fixing this though I'm still having an issue. In `assemble_maj` `pval` needs to be a numeric value (i.e. Float64 or such). The previous code was evaluating the full rate expression to generate this numeric value from the parameter vector. On the latest master and after removing the extra `ptoid` argument, for me the new code seems to return an MT expression for `pval`, instead of substituting in the numeric values from the parameter vector and then evaluating the numeric value of `sr`.

**Comment by shashi** [comment]
@isaacsas oops, sorry about that! do you think you can give me a small test that I can try to fix?

**Comment by shashi** [comment]
Ah ok, I just saw your comment on that PR, thanks!

**Comment by isaacsas** [comment]
Awesome! If you just go from that PR you can run the JumpSystem tests to check what’s going on.

**Pull request merged**